mmf/datasets/builders/visual7w/dataset.py

```diff
@@ -0,0 +1,109 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+
+import torch
+import numpy as np
+from mmf.common.sample import Sample
+from mmf.datasets.mmf_dataset import MMFDataset
+
+
+class Visual7WDataset(MMFDataset):
+    def __init__(self, config, *args, dataset_name="visual7w", **kwargs):
+        super().__init__(dataset_name, config, *args, **kwargs)
+
+    def iou(self, anchors, gt_boxes):
+        """
+        anchors: (N, 4) ndarray of float
+        gt_boxes: (K, 4) ndarray of float
+        overlaps: (N, K) ndarray of overlap between boxes and query_boxes
+        """
+        N = anchors.size(0)
+        K = gt_boxes.size(0)
+
+        gt_boxes_area = (
+            (gt_boxes[:, 2] - gt_boxes[:, 0] + 1) * (gt_boxes[:, 3] - gt_boxes[:, 1] + 1)
+        ).view(1, K)
+
+        anchors_area = (
+            (anchors[:, 2] - anchors[:, 0] + 1) * (anchors[:, 3] - anchors[:, 1] + 1)
+        ).view(N, 1)
+
+        boxes = anchors.view(N, 1, 4).expand(N, K, 4)
+        query_boxes = gt_boxes.view(1, K, 4).expand(N, K, 4)
+
+        iw = (
+            torch.min(boxes[:, :, 2], query_boxes[:, :, 2])
+            - torch.max(boxes[:, :, 0], query_boxes[:, :, 0])
+            + 1
+        )
+        iw[iw < 0] = 0
+
+        ih = (
+            torch.min(boxes[:, :, 3], query_boxes[:, :, 3])
+            - torch.max(boxes[:, :, 1], query_boxes[:, :, 1])
+            + 1
+        )
+        ih[ih < 0] = 0
+
+        ua = anchors_area + gt_boxes_area - (iw * ih)
+        overlaps = iw * ih / ua
+
+        return overlaps
+
+    def __getitem__(self, idx):
+        sample_info = self.annotation_db[idx]
+        features_data = self.features_db[idx]
+        multiple_choice_idx = torch.from_numpy(np.array(sample_info["mc_idx"]))
+
+        boxes = features_data['image_info_0']['bbox']
+        gt_boxes = \
+            features_data['image_info_1']['bbox'][multiple_choice_idx, :]
+        features = features_data['image_feature_0']
+        gt_features = \
+            features_data['image_feature_1'][multiple_choice_idx, :]
+
+        num_boxes = features.shape[0]
+        g_feat = torch.sum(features, 0) / num_boxes
+        g_feat = g_feat.unsqueeze(0)
+        image_w = features_data['image_info_0']['image_width']
+        image_h = features_data['image_info_0']['image_height']
+
+        g_location_ori = np.array(
+            [[0, 0, image_w, image_h]]
+        ).astype(np.float32)
+
+        features = torch.cat(
+            (g_feat, features), 0
+        )
+        boxes = np.concatenate((g_location_ori, boxes), axis=0)
```

Comment:
Convert these to torch tensor and then use torch methods to do all of the modifications. In general, always prefer torch methods over numpy methods for consistency.

Reply:
Done